### ISSUE TITLE
Add mechanism for generating public query functions and types

### ISSUE THREAD

**Issue by fmoor** [edgedb/edgedb-go#287]
Currently `edgeql-go` generates private query functions and return types. The user can export these types but it requires a few lines of code for each query. There should be a way to make the generator generate public instead of private functions and types.

Maybe this should be configurable on a per query basis?

**Comment by chirino** [comment]
I would default to being public, if a user really wants to keep it private he can place it under an internal directory so that it's public to external packages.

**Comment by fmoor** [comment]
> I would default to being public, if a user really wants to keep it private he can place it under an internal directory so that it's public to external packages.

Public by default seems reasonable, but I think there should still be a way to make them private to a package. In a large enough project it is reasonable to want to restrict some things more than just `internal/`.

**Comment by GenarLoya** [comment]
Hello, a little question, @fmoor, triying use the flag -pubfuncs not working, I suppose is not in the current release, can you help me to pass private to public, without modify generated files?

**Comment by fmoor** [comment]
> Hello, a little question, @fmoor, triying use the flag -pubfuncs not working, I suppose is not in the current release, can you help me to pass private to public, without modify generated files?

This is not implemented yet. You can achieve the same thing by putting this in a non generated file in the same package.
```go
var MyQueryFunc = myQueryFunc
```

I'll try to get this implemented soon.

**Comment by mdthompson-helium** [comment]
pubtypes generates a public struct but its fields are all private. Was it intentional to keep these private as well?

**Comment by fmoor** [comment]
> pubtypes generates a public struct but its fields are all private. Was it intentional to keep these private as well?

Yes. In EdgeQL it is possible to have both a `Name` and a `name` property in a shape/type. Because of this the default behavior is to preserved field names exactly as you spell them in your query. You can change this by either using the `-mixedcaps` flag or by using mixed caps names in your query. For example:

```edgeql
select schema::Function {
  Name := .name,
  Language := .language,
  Params := .params {
    Name := .name,
    Default := .default,
  }
}
```

You could also define your types with mixed caps names so you don't need to rename every thing in every query.

**Comment by fmoor** [comment]
> Is best practice to name those fields so they will be generated so they are public or generate the return type by default with public fields if you use pubtypes?

I don't think there is an established best practice here.

**Comment by mdthompson-helium** [comment]
> Yes. In EdgeQL it is possible to have both a `Name` and a `name` property in a shape/type. Because of this the default behavior is to preserved field names exactly as you spell them in your query. You can change this by either using the `-mixedcaps` flag or by using mixed caps names in your query.

Thanks, that helped clarify the behavior

**Comment by fmoor** [comment]
The new `-pubfucs` and `-pubtypes` options for the code generator are available in [v0.17.2](https://github.com/edgedb/edgedb-go/releases/tag/v0.17.2)
